### PR TITLE
Add efficientNet to image classification.

### PR DESCRIPTION
**Pull request by LielinJiang**
Add efficientNet to image classification.

### CODE REVIEW

File: PaddleCV/image_classification/models/efficientnet.py [hunk 1]
```diff
@@ -0,0 +1,470 @@
...
+import re
+import math
+import copy
+from .layers import conv2d, init_batch_norm_layer, init_fc_layer
+
+
+__all__ = ['EfficientNet', 'EfficientNetB0', 'EfficientNetB1', 'EfficientNetB2', 'EfficientNetB3', 'EfficientNetB4',
+           'EfficientNetB5', 'EfficientNetB6', 'EfficientNetB7']
+
+train_parameters = {
```
Comment (shippingwang) [inline PaddleCV/image_classification/models/efficientnet.py]: train_parameters is useless, should remove it.

File: PaddleCV/image_classification/models/efficientnet.py [hunk 2]
```diff
@@ -0,0 +1,470 @@
...
+
+def round_repeats(repeats, global_params):
+    """ Round number of filters based on depth multiplier. """
+    multiplier = global_params.depth_coefficient
+    if not multiplier:
+        return repeats
+    return int(math.ceil(multiplier * repeats))
+
+
+def swish(x):
```
Comment (shippingwang) [inline PaddleCV/image_classification/models/efficientnet.py]: same as fluid.layers.swish ?

File: PaddleCV/image_classification/models/efficientnet.py [hunk 3]
```diff
@@ -0,0 +1,470 @@
...
+        'efficientnet-b6': (1.8, 2.6, 528, 0.5),
+        'efficientnet-b7': (2.0, 3.1, 600, 0.5),
+    }
+    return params_dict[model_name]
+
+
+def efficientnet(width_coefficient=None, depth_coefficient=None,
+                 dropout_rate=0.2, drop_connect_rate=0.2):
+
+    blocks_args = [
```
Comment (shippingwang) [inline PaddleCV/image_classification/models/efficientnet.py:47]: what are these codes meaning?

File: PaddleCV/image_classification/models/layers.py [hunk 4]
```diff
@@ -0,0 +1,222 @@
...
+def cal_padding(img_size, stride, filter_size, dilation=1):
+    """Calculate padding size."""
+    if img_size % stride == 0:
+        out_size = max(filter_size - stride, 0)
+    else:
+        out_size = max(filter_size - (img_size % stride), 0)
+    return out_size // 2, out_size - out_size // 2
+
+def init_batch_norm_layer(name="batch_norm"):
+    param_attr = fluid.ParamAttr(
```
Comment (shippingwang) [inline PaddleCV/image_classification/models/layers.py:71]: In other models, params' weights are named as "_weights" and others: _offset , _scale , _mean , _variance

File: PaddleCV/image_classification/utils/optimizer.py [hunk 5]
```diff
@@ -67,6 +68,33 @@ def cosine_decay_with_warmup(learning_rate, step_each_epoch, epochs=120):
...
+    dtype='float32',
+    persistable=True,
+    name="learning_rate")
+
+    warmup_epoch = fluid.layers.fill_constant(
+        shape=[1], dtype='float32', value=float(warm_up_epoch), force_cpu=True)
+
+    with init_on_cpu():
+        epoch = ops.floor(global_step / step_each_epoch)
+        with control_flow.Switch() as switch:
```
Comment (shippingwang) [inline PaddleCV/image_classification/utils/optimizer.py]: There exists another "fluid.layers.control_flow" in "cosine_decay_with_warmup", feel free to refine it

File: PaddleCV/image_classification/utils/optimizer.py [hunk 6]
```diff
@@ -67,6 +68,33 @@ def cosine_decay_with_warmup(learning_rate, step_each_epoch, epochs=120):
...
+            with switch.case(epoch < warmup_epoch):
+                decayed_lr = learning_rate * (global_step / (step_each_epoch * warmup_epoch))
+                fluid.layers.tensor.assign(input=decayed_lr, output=lr)
+            with switch.default():
+                div_res = (global_step - warmup_epoch * step_each_epoch) / decay_steps
+                div_res = ops.floor(div_res)
+                decayed_lr = learning_rate * (decay_rate ** div_res)
+                fluid.layers.tensor.assign(input=decayed_lr, output=lr)
```
Comment (shippingwang) [inline PaddleCV/image_classification/utils/optimizer.py]: use short name "layers.assign()" ?

File: PaddleCV/image_classification/utils/optimizer.py [hunk 7]
```diff
@@ -123,8 +151,11 @@ def __init__(self, args):
         self.momentum_rate = args.momentum_rate
         self.step_epochs = args.step_epochs
         self.num_epochs = args.num_epochs
-
+        self.warm_up_epochs = args.warm_up_epochs
+        self.decay_steps = args.decay_steps
```
Comment (shippingwang) [inline PaddleCV/image_classification/utils/optimizer.py]: confused with self.step_epochs and self.warm_up_epochs, self.decay_steps

File: PaddleCV/image_classification/utils/utility.py [hunk 8]
```diff
@@ -115,6 +117,14 @@ def parse_args():
     add_arg('reader_thread',            int,    8,                      "The number of multi thread reader")
     add_arg('reader_buf_size',          int,    2048,                   "The buf size of multi thread reader")
     add_arg('interpolation',            int,    None,                   "The interpolation mode")
+    add_arg('use_ema',                  bool,   False,                  "Whether to use ExponentialMovingAverage.")
```
Comment (shippingwang) [inline PaddleCV/image_classification/utils/utility.py]: classify these arguments?

File: PaddleCV/image_classification/utils/utility.py [hunk 9]
```diff
@@ -115,6 +117,14 @@ def parse_args():
...
+    add_arg('use_aa',                   bool,   False,                  "Whether to use auto augment")
+    add_arg('ema_decay',                float,  0.9999,                 "The value of ema decay rate")
+    add_arg('warm_up_epochs',           float,  5.0,                    "The value of warm up epochs")
```
Comment (shippingwang) [inline PaddleCV/image_classification/utils/utility.py]: int?

**Review by shippingwang — CHANGES_REQUESTED** [review]

**Comment by shippingwang** [comment]
and add parameters explanation in readme & readme_en, if pretrained model is supported, update documents

File: PaddleCV/image_classification/models/efficientnet.py (hunk 1 above)
Comment (LielinJiang) [reply PaddleCV/image_classification/models/efficientnet.py]: removed.

File: PaddleCV/image_classification/models/efficientnet.py (hunk 2 above)
Comment (LielinJiang) [reply PaddleCV/image_classification/models/efficientnet.py]: yes, removed.

File: PaddleCV/image_classification/models/efficientnet.py (hunk 3 above)
Comment (LielinJiang) [reply PaddleCV/image_classification/models/efficientnet.py:47]: added comment.

File: PaddleCV/image_classification/models/layers.py (hunk 4 above)
Comment (LielinJiang) [reply PaddleCV/image_classification/models/layers.py:71]: updated.

File: PaddleCV/image_classification/utils/optimizer.py (hunk 5 above)
Comment (LielinJiang) [reply PaddleCV/image_classification/utils/optimizer.py]: updated.

File: PaddleCV/image_classification/utils/optimizer.py (hunk 6 above)
Comment (LielinJiang) [reply PaddleCV/image_classification/utils/optimizer.py]: updated.

File: PaddleCV/image_classification/utils/optimizer.py (hunk 7 above)
Comment (LielinJiang) [reply PaddleCV/image_classification/utils/optimizer.py]: change decay_steps to decay_epochs, and add comment in readme.

File: PaddleCV/image_classification/utils/utility.py (hunk 9 above)
Comment (LielinJiang) [reply PaddleCV/image_classification/utils/utility.py]: warm up epochs can be float, if do not want to warm up one epoch, like in yolo v3.

File: PaddleCV/image_classification/utils/utility.py (hunk 8 above)
Comment (LielinJiang) [reply PaddleCV/image_classification/utils/utility.py]: > and add parameters explanation in readme & readme_en, if pretrained model is supported, update documents

added.

**Comment by LielinJiang** [comment]
> and add parameters explanation in readme & readme_en, if pretrained model is supported, update documents

added.

File: PaddleCV/image_classification/utils/utility.py (hunk 8 above)
Comment (LielinJiang) [reply PaddleCV/image_classification/utils/utility.py]: updated

**Review by shippingwang — APPROVED** [review]
Good job!

**Pull request merged**